File: apina-java/src/test/java/fi/evident/apina/java/model/JavaAnnotationTest.java
package fi.evident.apina.java.model;

import org.junit.Test;

import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class JavaAnnotationTest {

    @Test
    public void toStringWithoutParameters() {
        JavaAnnotation annotation = newAnnotation("foo.bar.Baz");
        assertThat(annotation.toString(), is("@foo.bar.Baz"));
    }

    @Test
    public void toStringWithJustValue() {
        JavaAnnotation annotation = newAnnotation("foo.bar.Baz");
        annotation.setAttribute("value", 42);

        assertThat(annotation.toString(), is("@foo.bar.Baz(42)"));
    }

    @Test
    public void toStringWithNonValueAttribute() {
        JavaAnnotation annotation = newAnnotation("foo.bar.Baz");
        annotation.setAttribute("foo", 42);

        assertThat(annotation.toString(), is("@foo.bar.Baz(foo=42)"));
    }

    @Test
    public void toStringWithMultipleAttributes() {
        JavaAnnotation annotation = newAnnotation("foo.bar.Baz");
        annotation.setAttribute("value", 1);
        annotation.setAttribute("foo", 2);
        annotation.setAttribute("bar", 3);

        assertThat(annotation.toString(), is("@foo.bar.Baz(value=1, foo=2, bar=3)"));
    }

    @Test
    public void arrayValues() {
        JavaAnnotation annotation = newAnnotation("foo.bar.Baz");
        annotation.setAttribute("array", new Object[] { 1, 2, 3 });
        annotation.setAttribute("single", 1);

        assertThat(annotation.getAttributeValues("array"), is(asList(1, 2, 3)));
        assertThat(annotation.getAttributeValues("single"), is(singletonList(1)));
        assertThat(annotation.getAttributeValues("nonexistent"), is(emptyList()));
    }

    private static JavaAnnotation newAnnotation(String name) {
        return new JavaAnnotation(new QualifiedName(name));
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/java/model/JavaAnnotationTest.kt
package fi.evident.apina.java.model

import fi.evident.apina.java.model.type.JavaType.Basic
import org.junit.jupiter.api.Test
import java.util.Collections.emptyList
import kotlin.test.assertEquals

class JavaAnnotationTest {

    @Test
    fun `toString without parameters`() {
        val annotation = newAnnotation("foo.bar.Baz")
        assertEquals("@foo.bar.Baz", annotation.toString())
    }

    @Test
    fun `toString with just value`() {
        val annotation = newAnnotation("foo.bar.Baz")
        annotation.setAttribute("value", 42)

        assertEquals("@foo.bar.Baz(42)", annotation.toString())
    }

    @Test
    fun `unwrap array arguments if necessary`() {
        val annotation = newAnnotation("foo.bar.Baz")
        annotation.setAttribute("value", arrayOf<Any>("foo"))

        assertEquals("foo", annotation.getAttribute<String>("value"))
    }

    @Test
    fun toStringWithNonValueAttribute() {
        val annotation = newAnnotation("foo.bar.Baz")
        annotation.setAttribute("foo", 42)

        assertEquals("@foo.bar.Baz(foo=42)", annotation.toString())
    }

    @Test
    fun `toString with multiple attributes`() {
        val annotation = newAnnotation("foo.bar.Baz")
        annotation.setAttribute("value", 1)
        annotation.setAttribute("foo", 2)
        annotation.setAttribute("bar", 3)

        assertEquals("@foo.bar.Baz(value=1, foo=2, bar=3)", annotation.toString())
    }

    @Test
    fun `array values`() {
        val annotation = newAnnotation("foo.bar.Baz")
        annotation.setAttribute("array", arrayOf<Any>(1, 2, 3))
        annotation.setAttribute("single", 1)

        assertEquals(listOf(1, 2, 3), annotation.getAttributeValues("array"))
        assertEquals(listOf(1), annotation.getAttributeValues("single"))
        assertEquals(emptyList(), annotation.getAttributeValues("nonexistent"))
    }

    private fun newAnnotation(name: String) = JavaAnnotation(Basic(name))
}
